[NFC] Never synthesize @lvalue in typing judgements

CodaFi opened the pull request:
When in a checking context, there is no need (and no sense) in synthesizing `@lvalue` instead of checking the underlying object type.

<strike>The locators for this typing rule and the other parameter binding one below it are still lying, but probably because there's really no good way to indicate that you're an argument to a function/operator without also lying about the index of the argument - which matchTypes has no way of knowing.</strike>

CodaFi commented:
@swift-ci please test

CodaFi commented:
@swift-ci please test source compatibility

xedin reviewed (commented):
I think this is reasonable, but it looks like it might require special handling for mutable keypaths?

CodaFi commented:
Aha, well I'm going to move the @lvalue synthesis to CSGen where it belongs.

CodaFi committed:
[NFC] Remove synthesis of @lvalue in typing judgement

Only two more of these exist
1) CSGen's formation of @lvalue -> inout conversion constraints
2) WritableKeyPath's checking rules

Both can be eliminated but will require more refactoring.

CodaFi commented:
LLDB test timed out

@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please test source compatibility

CodaFi commented:
@swift-ci please test source compatibility

CodaFi commented:
⛵️
